cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,231 @@
+// limitations under the License.
+package main
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
+	"github.com/projectcalico/libcalico-go/lib/apis/v3"
+	bapi "github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/k8s"
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/libcalico-go/lib/ipam"
+	"github.com/projectcalico/libcalico-go/lib/logutils"
+	cnet "github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/options"
+)
+
+const (
+	ipAllocPath = "/var/lib/cni/networks/k8s-pod-network"
+)
+
+func main() {
+	// nodename associates IPs to this node.
+	nodename := os.Getenv("KUBERNETES_NODE_NAME")
+	logCtxt := log.WithField("node", nodename)
+
+	if msg, err := Main(nodename); err != nil {
+		logCtxt.WithError(err).Fatal(msg)
+	}
+}
+
+func Main(nodename string) (msg string, err error) {
+	// Declare variables to avoid shadowing err.
+	var (
+		cfg          *apiconfig.CalicoAPIConfig
+		calicoClient client.Interface
+		clusterInfo  *v3.ClusterInformation
+		node         *v3.Node
+		files        []os.FileInfo
+		ip           *cnet.IP
+		b            []byte
+		calicoDS     *v1beta1.DaemonSet
+		podList      *corev1.PodList
+	)
+
+	// Set up logging formatting.
+	log.SetFormatter(&logutils.Formatter{})
+
+	// Install a hook that adds file/line no information.
+	log.AddHook(&logutils.ContextHook{})
+
+	// Check to see if the system is still using host-local
+	// by checking the existence of the path.
+	if _, err = os.Stat(ipAllocPath); err != nil && os.IsNotExist(err) {
+		log.Info("no migration necessary, successfully exiting...")
+		return
+	}
+
+	//
+	// Initialize necessary variables.
+
+	// ctxt for network calls
+	ctxt := context.Background()
+
+	// calicoClient makes IPAM calls.
+	cfg, err = apiconfig.LoadClientConfig("")
+	if err != nil {
+		msg = "failed to load api client config"
+		return
+	}
+	cfg.Spec.DatastoreType = apiconfig.Kubernetes
+	if calicoClient, err = client.New(*cfg); err != nil {
+		msg = "failed to initialize api client"
+		return
+	}
+
+	// k8sClient directly calls the k8s apiserver.
+	type accessor interface {
+		Backend() bapi.Client
+	}
+	k8sClient := calicoClient.(accessor).Backend().(*k8s.KubeClient).ClientSet
+
+	// Disable cni
+	if clusterInfo, err = calicoClient.ClusterInformation().Get(ctxt, "default", options.GetOptions{}); err != nil {
+		msg = "failed to fetch cluster information"
+		return
+	}
+	if *clusterInfo.Spec.DatastoreReady {
+		*clusterInfo.Spec.DatastoreReady = false
+		if clusterInfo, err = calicoClient.ClusterInformation().Update(ctxt, clusterInfo, options.SetOptions{}); err != nil {
+			msg = "failed to disable cluster"
+			return
+		}
+	}
```

Comment:
I think we need a reasonable sleep here to allow any in-progress CNI executions to finish - maybe 5 seconds?

Reply:
@briansan I don't see a sleep... probably should be below the part where we delete the binaries.